src/server/pkg/storage/fileset/storage.go

```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
 	})
 }
 
+// SetTTL sets the time-to-live for the prefix. If the prefix contains multiple filesets
+// the ttl will be set on all of them and the soonest expiration time will be returned.
+// if no fileset is found SetTTL returns ErrNoFileSetFound
+func (s *Storage) SetTTL(ctx context.Context, prefix string, ttl time.Duration) (time.Time, error) {
+	// TODO: after we move the semantic paths to postrges the whole prefix can be set atomically.
+	var expiresAt *time.Time
+	if err := s.objC.Walk(ctx, applyPrefix(prefix), func(p string) error {
+		t, err := s.chunks.RenewReference(ctx, p, ttl)
+		if err != nil {
+			return err
+		}
+		if expiresAt == nil || t.Before(*expiresAt) {
+			expiresAt = &t
+		}
+		return nil
+	}); err != nil {
+		return time.Time{}, err
+	}
+	if expiresAt == nil {
+		return time.Time{}, ErrNoFileSetFound
+	}
+	return *expiresAt, nil
+}
+
+// RenewDuring ensures that none of the filesets under prefix expire by repeatedly calling SetTTL in a separate goroutine
+// until cb returns.
+// RenewDuring attempts to delete the contents of prefix after cb has returned
+func (s *Storage) RenewDuring(ctx context.Context, prefix string, cb func(context.Context) error) error {
+	const defaultTTL = 10 * time.Minute
+	eg, ctx2 := errgroup.WithContext(ctx)
+	ctx2, cancel := context.WithCancel(ctx2)
+	eg.Go(func() error {
+		var expiresAt time.Time
+		var err error
+		for {
+			expiresAt, err = s.SetTTL(ctx2, prefix, defaultTTL)
```

Comment:
I don't think this is the algorithm we want for two reasons: it is not as simple as renewing at a fixed period, and it will actually result with more requests when high latency is encountered which is probably not the appropriate behavior for the system as a whole. I think we should look at this as more of a cooperative process, and error out appropriately when high latency is encountered (so a renewal will eventually fail in this case).

Reply:
This is changed to a simple ticker. And due to the new semantics of `WithRenewal` we now detect slowness with not exist errors.